### PR TITLE
Finnish address search in the search bar

### LINKED ISSUE #1695: Feature request: geocoding & address search (Finland)
As a user I would like to be able to search for addresses with QField by using a third party API. Currently [the search works with layers](https://qfield.org/docs/et/old-doc/development/index.html), but not with external API's. In Finland there is an open geocoding API that could be used for this purpose: https://digitransit.fi/en/developers/apis/2-geocoding-api/

Addresses could be searched with [address lookup](https://digitransit.fi/en/developers/apis/2-geocoding-api/address-lookup/). The API has a rate limit of 10 requests per second which should very well be sufficient to cover the needs of Finnish QField users 😊. We have also done [a QGIS plugin that uses this API](https://github.com/GispoCoding/QGISDigitransitGeocoding) if that would be helpful in any way in the implementation.

Geocoding has already been mentioned as one of the identified development topics[ in the docs. ](https://qfield.org/docs/et/old-doc/development/index.html) and a discussion about adding this functionality has been already started with @m-kuhn 

One open question is that how this could be activated for only users in Finland. Also as the API is pretty generic and based on [Pelias](https://github.com/pelias/documentation), I would hope implementing this could benefit also other locations. I am open for discussion regarding this implementation.

### PR DESCRIPTION
**Pull request by nirvn**
This PR implements Finnish address search capability within QField's top search bar.

Showcase MP4:

https://user-images.githubusercontent.com/1728657/109791240-cafb0180-7c44-11eb-8f97-d9787336b9bf.mp4

Search results are remembered when re-opening the search bar after you clicked on one result, which helps navigate through the returned locations.

As part of the implementation of this feature, a nice pair of Pelias geocoder and locator filter classes were created, using QGIS' base classes. 

The Finnish geocoder will auto enable its non-prefix use in the search bar whenever the user switches the position/location/GPS button on and overlaps with Finland's bounding box -- if the user subsequently switch it off, QField will be respectful enough not to turn it back on.

The position check is only done _when the user toggles the position/location/GPS button on_, this way, we insure that we're using such data when the user has explicitly expressed he/she is OK fetching that data to begin with.

Fixes #1695

### CODE REVIEW

**Comment by nirvn** [comment]
(This PR implements the requested feature #1695)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 1f0f782540fa28fc8da91028162d9aaa35dd4f66:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-1fdcc854318c213a6ff8609d0e27b369557eaf40-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-1fdcc854318c213a6ff8609d0e27b369557eaf40-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-1fdcc854318c213a6ff8609d0e27b369557eaf40-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-1fdcc854318c213a6ff8609d0e27b369557eaf40-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 17b1018d3134a8df5c9dde07638efcf6e29e4ffa:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-b7b7265f367af74c1ca7927f3a4c984cef2d8704-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-b7b7265f367af74c1ca7927f3a4c984cef2d8704-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-b7b7265f367af74c1ca7927f3a4c984cef2d8704-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-b7b7265f367af74c1ca7927f3a4c984cef2d8704-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 6c79c3cfc24436e35c743e1419d6a209a64fa5dd:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-c19f1efab2d69c157fc7155bc534cd1936286f40-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-c19f1efab2d69c157fc7155bc534cd1936286f40-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-c19f1efab2d69c157fc7155bc534cd1936286f40-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-c19f1efab2d69c157fc7155bc534cd1936286f40-x86.apk)

File: src/core/peliasgeocoder.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+#include <QDateTime>
+#include <QUrl>
+#include <QUrlQuery>
+#include <QMutex>
+#include <QNetworkRequest>
+#include <QJsonDocument>
+#include <QJsonArray>
+
+QMutex PeliasGeocoder::sMutex;
+QMap< QUrl, QList< QgsGeocoderResult > > PeliasGeocoder::sCachedResults;
```
Comment (nyalldawson) [inline src/core/peliasgeocoder.cpp:31]: Warning - you should use qglobalstatic on this or you'll get crashes on exit

Comment (nirvn) [reply src/core/peliasgeocoder.cpp]: @nyalldawson , thanks for the tip, adjusted.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 64e4f4cea19d1ab2c413477a284d588287a00d71:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-846224dbd0a3400f53b3a311973858412911fc57-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-846224dbd0a3400f53b3a311973858412911fc57-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-846224dbd0a3400f53b3a311973858412911fc57-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-846224dbd0a3400f53b3a311973858412911fc57-x86.apk)

**Comment by nirvn** [comment]
Alright, the UI is in. It looks like this:

https://user-images.githubusercontent.com/1728657/109949157-a0c14680-7d0d-11eb-87bb-b65b9ac9280f.mp4

The design relies on a model to make it future-proof. This allows us to expose all registered search/locator filters to the user, and given them the power to decide which filter should be on all the time or only triggered via its prefix.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for d8c35771a2ffff1f7096b2edaab1f14c4fa8a462:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-311cc6ddc067b6ea8d0a316164a8d99d5ee28457-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-311cc6ddc067b6ea8d0a316164a8d99d5ee28457-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-311cc6ddc067b6ea8d0a316164a8d99d5ee28457-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-311cc6ddc067b6ea8d0a316164a8d99d5ee28457-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 60f34f2fb1bad661dfd2b0f51246ee673937bc71:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0270f6383a28924e967bb81bf64ba3988e17213b-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0270f6383a28924e967bb81bf64ba3988e17213b-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0270f6383a28924e967bb81bf64ba3988e17213b-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0270f6383a28924e967bb81bf64ba3988e17213b-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 32ca0f7999a030d77e49a5a25594c5ad34fa91ff:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-551227f9320eac7205b73c1f6e3dbd4ef345be94-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-551227f9320eac7205b73c1f6e3dbd4ef345be94-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-551227f9320eac7205b73c1f6e3dbd4ef345be94-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-551227f9320eac7205b73c1f6e3dbd4ef345be94-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e2026c6fad3a043d5b1cb886cf927066bb3667e9:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-ff492ab49914d928e4d6c1144e052f017657ee49-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-ff492ab49914d928e4d6c1144e052f017657ee49-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-ff492ab49914d928e4d6c1144e052f017657ee49-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-ff492ab49914d928e4d6c1144e052f017657ee49-x86.apk)

**Comment by nirvn** [comment]
The last piece of the puzzle is in: the Finnish geocoder will auto enable its non-prefix use in the search bar whenever the user switches the position/location/GPS button on and overlaps with Finland's bounding box :tada: -- if the user subsequently switch it off, QField will be respectful enough not to turn it back on.

@m-kuhn , that's pretty much it, a review is most appreciated and frankly needed :wink:

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 249ace2191684b931f3bef2765783320c217e8e9:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0ed310ab643ff21e51ec16981203147e9a7becf3-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0ed310ab643ff21e51ec16981203147e9a7becf3-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0ed310ab643ff21e51ec16981203147e9a7becf3-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-0ed310ab643ff21e51ec16981203147e9a7becf3-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 9e42be85894931c2002f1eb98c203f8458e6a61d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-811250aa9fd60c9a6e7d00d1179c4658fcf62280-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-811250aa9fd60c9a6e7d00d1179c4658fcf62280-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-811250aa9fd60c9a6e7d00d1179c4658fcf62280-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-811250aa9fd60c9a6e7d00d1179c4658fcf62280-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 924dd92d68bb95cb595de59623dcb33b90bf1355:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-270a3f664778bd23e07f6482dac5b3af2ad864e7-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-270a3f664778bd23e07f6482dac5b3af2ad864e7-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-270a3f664778bd23e07f6482dac5b3af2ad864e7-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-270a3f664778bd23e07f6482dac5b3af2ad864e7-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ab111d60a5a7e7abcb05f785a60908ef19bfaee0:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-f0e2788f6971a09acfb297270037e31b93aef404-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-f0e2788f6971a09acfb297270037e31b93aef404-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-f0e2788f6971a09acfb297270037e31b93aef404-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-f0e2788f6971a09acfb297270037e31b93aef404-x86.apk)

**Review by m-kuhn — APPROVED** [review]
🔍 🗺️ 🧖

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 1b6b2074beaed5f81d34fe0ee859c73256d599d0:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-8eacee7312b88b0c50466f5684f6bb2613f82907-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-8eacee7312b88b0c50466f5684f6bb2613f82907-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-8eacee7312b88b0c50466f5684f6bb2613f82907-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1739-8eacee7312b88b0c50466f5684f6bb2613f82907-x86.apk)

**Pull request merged**